pkg/member/member_add.go

```diff
@@ -0,0 +1,264 @@
+package member
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
+	"gopkg.in/yaml.v2"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// retryPeriod is the peroid after which an operation is retried
+	retryPeriod = 5 * time.Second
+
+	// etcdTimeout is timeout for etcd operations
+	etcdTimeout = 10 * time.Second
+)
+
+// AddMember interface defines the functionalities needed to add a member to the etcd cluster
+type AddMember interface {
+	// AddMemberAsLearner add a new member as a learner to the etcd cluster
+	AddMemberAsLearner(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+	IsMemberInCluster(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
+	PromoteMember(context.Context, *logrus.Entry, *brtypes.EtcdConnectionConfig)
+}
+
+// NewMember holds the configuration for the mechanism of adding a new member to the cluster.
+type NewMember struct {
+	clientFactory client.Factory
+}
+
+// NewMemberConfig returns new ExponentialBackoff.
+func NewMemberConfig(etcdConnConfig *brtypes.EtcdConnectionConfig) *NewMember {
+	etcdConn := *etcdConnConfig
+	svcEndpoint, err := miscellaneous.GetEtcdSvcEndpoint()
+	if svcEndpoint == "" || err != nil {
+		svcEndpoint = "http://127.0.0.1:2379"
+	}
+	etcdConn.Endpoints = []string{svcEndpoint}
+	clientFactory := etcdutil.NewFactory(etcdConn)
+	return &NewMember{
+		clientFactory: clientFactory,
+	}
+}
+
+// AddMemberAsLearner add a member as a learner to the etcd cluster
+func (m *NewMember) AddMemberAsLearner(logger *logrus.Logger) error {
+	//Add member as learner to cluster
+	logger.Info("Starting Add Member")
+	memberURL, err := getMemberURL(logger)
+	if memberURL == "" || err != nil {
+		logger.Info("Error fetching etcd member URL")
+		return fmt.Errorf("Could not fetch member URL : %v", err)
+	}
+
+	cli, err := m.clientFactory.NewCluster()
+	defer cli.Close()
+	if err != nil {
+		logger.Info("Error creating etcd client to add member")
+		return fmt.Errorf("failed to build etcd cluster client : %v", err)
+	}
+
+	memAddCtx, cancel := getContext(context.TODO(), etcdTimeout)
```

Comment:
this method is not so useful and therefore it should be removed. Also passing `context.TODO` instead of taking parent context that should have been passed to the method needs to be corrected.